BJS2-53971: Стать ментором

ShishovNikita opened the pull request:
Добавлены основные операции возможностей быть ментором:

- Получить всех менти одного пользователя
- Получить всех менторов одного пользователя
- Удалить менти
- Удалить ментора

Добавлены тесты на сервис и контроллер

ShishovNikita committed:
BJS2-53971: Стать ментором

ShishovNikita committed:
BJS2-53971: Стать ментором

ShishovNikita committed:
Merge remote-tracking branch 'origin/BJS2-53971/Стать-ментором' into BJS2-53971/become-a-mentor

ShishovNikita committed:
Переименован класс сервиса

sstani-bgv commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,63 @@
+    public List<UserMentorshipDto> getMentees(long userId) {
+        User mentor = mentorshipRepository.findById(userId).orElseThrow(
+                () -> new NoSuchElementException("Не существует пользователя с ID: " + userId));
+
```
Можно декомпозировать в отдельный метод getUser

sstani-bgv commented on src/test/java/school/faang/user_service/service/MentorshipRelationServiceTest.java:
```diff
@@ -0,0 +1,182 @@
+        assertUserNotFound(exception, userId);
+    }
+
+    private long arrange() {
```
Предлагаю более информативное название mockMentorshipRepositoryFindByID

sstani-bgv reviewed (approved):
Молодец, отличный код!

ShishovNikita committed:
Декомпозирован код в метод getUser
Изменено название метода assertUserNotFound -> mockMentorshipRepositoryFindByID

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+    public List<UserMentorshipDto> getMentees(long userId) {
+        User mentor = getUser(userId);
+
+        return mentor.getMentees().stream()
```
```Optional.ofNullable(mentor.getMentees()).orElseGet(ArrayList::new).stream()...```

Чтобы NPE не схватить

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+
+    private User getUser(long userId) {
+        return mentorshipRepository.findById(userId).orElseThrow(
+                () -> new NoSuchElementException("Не существует пользователя с ID: " + userId));
```
Заменить на EntityNotFoundException

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+    public List<UserMentorshipDto> getMentors(long userId) {
+        User mentee = getUser(userId);
+
+        return mentee.getMentors().stream()
```
Optional.ofNulable()

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+
+    public void deleteMentee(long menteeId, long mentorId) {
+        User mentor = mentorshipRepository.findById(mentorId).orElseThrow(() ->
+                new NoSuchElementException(String.format("Ментор с ID: %d не найден!!!", mentorId)));
```
EntityNotFoundException

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+                new NoSuchElementException(String.format("Ментор с ID: %d не найден!!!", mentorId)));
+
+        if (!mentor.getMentees().removeIf((mentee) -> mentee.getId().equals(menteeId))){
+            throw new NoSuchElementException(String.format("Менти с ID: %d не найден!!!", menteeId));
```
EntityNotFoundException

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        mentorshipRepository.save(mentor);
+        log.info("Менти с ID: {} из списка ментора с ID: {} удален.", menteeId, mentorId);
```
Лишний лог, можно убрать

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+
+    public void deleteMentor(long menteeId, long mentorId) {
+        User mentee = mentorshipRepository.findById(menteeId).orElseThrow(() ->
+                new NoSuchElementException(String.format("Менти с ID: %d не найден!!!", menteeId)));
```
EntityNotFoundException

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+                new NoSuchElementException(String.format("Менти с ID: %d не найден!!!", menteeId)));
+
+        if(!mentee.getMentors().removeIf((mentor) -> mentor.getId().equals(mentorId))){
+            throw new NoSuchElementException(String.format("Ментор с ID: %d не найден!!!", mentorId));
```
EntityNotFoundException

DmShevchuk commented on src/main/java/school/faang/user_service/service/MentorshipRelationService.java:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        mentorshipRepository.save(mentee);
+        log.info("Ментор с ID: {} из списка менти с ID: {} удален", mentorId, menteeId);
```
Лишний лог

ShishovNikita committed:
NoSuchElementException заменен на EntityNotFoundException
Добавлен Optional.ofNulable()
